spark-ui/src/reducers/SqlReducer.ts

```diff
@@ -399,6 +400,7 @@ export function updateSqlNodeMetrics(
 
   const notEffectedSqls = currentStore.sqls.filter((sql) => sql.id !== sqlId);
   const runningSql = runningSqls[0];
+  const graph = generateGraph(runningSql.edges, runningSql.nodes);
```

Comment:
This means we are generating the graph for every metric update cycle. This is not ideal.
We should cache the graph if there is no change.
Please at least add a todo comment to cache the graph.

Reply:
Yes, I thought about it, but my bad to forget the todo.
I think it should work with map with sqlId key easily.
But detecting actual changes is something different. We can do a check on node metrics before and after or compare their hashes. What do you think?